### PR TITLE
Simplify network analysis

### PR DESCRIPTION
**Pull request by ajlee21**
This PR simplifies how we are generating networks. The current steps to generate the network include performing a correlation analysis to generate a binary adjacency matrix, this adjacency matrix is then used to create a similarity matrix (Topological Overlap matrix) based on if genes are connected to the same set of genes, and finally hierarchical clustering on this similarity matrix to identify modules.

In order to make our approach more intuitive, we will simplify the steps used to generate the network by first performing the same correlation analysis and threshold based on some cutoff. But then apply clustering to these filtering correlation scores to identify modules

Changes in this PR include:
1. Add clustermap visualization of the raw correlation matrix and TOM matrix to determine the appropriate input: `1_correlation_analysis.ipynb` and `1a_get_network_communities_complex.ipynb`. The clustermap visualizations were saved offline due to their large sizes. To see examples of what they look like, see slides 2-6: https://docs.google.com/presentation/d/10Pr_oD5jL8GxCHLg1Qb16A6br31E2COgZIThjnj5ZyI/edit?usp=sharing

2. Add new notebook to perform clustering and extract module membership: `2_get_network_communities.ipynb`
3. Add Readme description.

The main takeaway from this PR is that:
1. It appears that the TOM matrix is not as sensitive since its setting nearby genes to have a score of 1.
2. Using just the correlation score is grouping genes into one large cluster. To dappen the overwhelming signal of highly correlated genes we will look into applying these two methods in the next PR: 
https://pubmed.ncbi.nlm.nih.gov/17724061/
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4768301/

### CODE REVIEW

File: networks/2_get_network_communities.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+# pa14_clustering = DBSCAN().fit(pa14_corr)
+# -
+
+# ### Hierarchal clustering
+# [Hierachal clustering](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.AgglomerativeClustering.html#sklearn.cluster.AgglomerativeClustering): Initially, each object is assigned to its own cluster and then the algorithm proceeds iteratively, at each stage joining the two most similar clusters (i.e. linkage distance is minimized), continuing until there is just a single cluster.
+#
+# * n_cluster: The number of clusters to find.
+# * linkage: Criterion used to determine distance between observations. 'average'=average distance of each observation in the two sets.
+#
+# * Note: It looks like this method tends to produce 1 very large cluster. To break this up we will iteratively apply hierarchal clustering on the largest cluster.
```
Comment (jjc2718) [inline networks/2_get_network_communities.py:78]: You could also try changing `n_clusters` - setting this to a larger number might give you a more even size distribution.

**Review by jjc2718 — APPROVED** [review]
Looks good to me! I definitely think what you're doing now is more straightforward, at least from my perspective. Also interesting (but not surprising) that binarizing the edges in the TOM method makes it less sensitive.

File: networks/2_get_network_communities.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+#
+# A cluster is a set of core samples that can be built by recursively taking a core sample, finding all of its neighbors that are core samples, finding all of their neighbors that are core samples, and so on. A cluster also has a set of non-core samples, which are samples that are neighbors of a core sample in the cluster but are not themselves core samples. Intuitively, these samples are on the fringes of a cluster.
+#
+# * We define a core sample as being a sample in the dataset such that there exist `min_samples` other samples within a distance of `eps`, which are defined as neighbors of the core sample.
+# * Using all default values
+#
+
+# +
+# Clustering using DBSCAN
+# pao1_clustering = DBSCAN().fit(pao1_corr)
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply networks/2_get_network_communities.py:68]: Yes definitely something I intend to do when I return to this step of the analysis. This PR was initially meant to apply those methods but got side tracked by the correlation visualization, which was a good step back to take. So I'll come back to this in a future PR

File: networks/2_get_network_communities.py (hunk 1 above)
Comment (ajlee21) [reply networks/2_get_network_communities.py:78]: This parameter confused me a bit. I wasn't sure if `n_clusters` referred to 1) the total number of final clusters or 2) if it was the number of clusters per merge (which I assumed was 2, but maybe that's not the case). I guess if it were (1) then I felt a little hesitant setting a number of total clusters to expect. One thing I was considering applying was this hdbscan that has fewer params to get around this too.

File: networks/1_correlation_analysis.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+# -
+
+pa14_corr.head()
+
+# ## Plot distribution of pairwise distances
+#
+# This will particularly help to inform the parameters we use for DBSCAN, which is density based
+
+# Get distribution of pairwise distances to determine a cutoff defining what a dense region should be
+f1 = sns.displot(pdist(pao1_corr))
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply networks/1_correlation_analysis.py:114]: This is a very good point. I was examining this distance as a way to inform the parameter I would choose for DBSCAN, so to determine how I am defining density. Since DBSCAN is actually looking at local regions of density I think adding this calculation would make sense. Thank you!

File: networks/1_correlation_analysis.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+# Random PAO1 genes
+random_pao1_ids = random.sample(list(pao1_compendium.columns), 4)
+sns.pairplot(pao1_compendium[random_pao1_ids])
+plt.suptitle("Random set of genes", y=1.05)
+
+# Co-operonic PAO1 genes
+# pao1_co_operonic_ids = ["PA0001", "PA0002", "PA0003", "PA0004"]
+# pao1_co_operonic_ids = ["PA0054","PA0055", "PA0056"]
+pao1_co_operonic_ids = ["PA1335", "PA1336", "PA1337"]
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply networks/1_correlation_analysis.py:72]: Yes, I would expect higher correlation between co-operonic genes compared to some random set of genes. So the scales of these two genes are different, which I think it what is causing the scatterplot to look flat -- if I truncate the samples to only include those up until the max expression of the other genes, the scatter plots look more linear. Something to keep in mind for when I do these correlations -- if I need to normalize for that the genes are within similar ranges.

File: networks/2_get_network_communities.py (hunk 1 above)
Comment (jjc2718) [reply networks/2_get_network_communities.py:82]: I think it's the total number of final clusters (i.e. where to cut the dendrogram to give the final output clusters).

I don't feel strongly about trying it, though - if you'd rather just run it recursively or use another method that's totally fine. You could also chat with Milton if you're trying to decide on a method, I know he's done a lot of clustering stuff in the past so he might know of some options that I'm not aware of.

File: networks/1_correlation_analysis.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jjc2718) [reply networks/1_correlation_analysis.py:75]: Good catch!
